### DISCUSSION TITLE
How to install mouse-actions on Ubuntu 24.04 or LinuxMint 22

### DISCUSSION THREAD

**Discussion by braikar** [jersou/mouse-actions#48] [General]
Hello,

I'm just sharing this as help for people who might want it on ubuntu 24.04 or linux-mint 22.

First of all! Thanks!! I used to have easystroke too, it's now deprecated (I just stuck with it much longer on a very outdated linuxmint...), I found your app and it seems that it is perfect as a replacement!

The amd64.deb (4.5.0) you just published is not ready for the dependencies of ubuntu 24.04/linux-mint 22.

But the app works, it just needs to be built manually (until there is, maybe a proper release.. But it doesn't seem to be the best time now to do it, tauri-v2 is still in beta and rust 1.8 might break small things soon it seems!).

(Btw. looking for replacements, I also ended-up on https://github.com/dkondor/wstroke which is very promising, just a port of easystroke, but using wayfire ("new" compiz) on wayland, so the edges handling are not required, because it's managed by wayfire. But I tried it, full on wayland, full on wayfire, is too unstable IMO for the moment! On my setups, it's blank screens almost everywhere... I'm sticking to x11 for the moment)

## GUI-Less version (works as-is):
Follow the instructions here: https://github.com/jersou/mouse-actions?tab=readme-ov-file#install--run
`cargo build --release`
The only think I had to find out it that there is a utility to create debian packages form cargo projects, so after the cargo build, use `cargo-deb` (That would be cool in the readme for debian specific help, not everyone might know that there is cargo-deb... I searched in the wrong direction for a while before figuring out there's this little handy tool!)
https://github.com/kornelski/cargo-deb
That's it, you'll have the gui-less package (after searching for all the dependencies missing..)
The package will be in `target/debian/mouse-actions_0.4.5-1_amd64.deb`

## With-GUI version (needs some changes):
Follow the instructions here again: https://github.com/jersou/mouse-actions?tab=readme-ov-file#install--run
`cd config-editor && npm i && npm run tauri-build`

Install all the dependencies that are needed and missing again with the help of the various error messages.

Then you will run into the unresolvable error (you can resolve all other ones by installing the right dependencies):
`The system library `javascriptcoregtk-4.0` required by crate `javascriptcore-rs-sys` was not found.`
Don't look for that dependency, it won't work...

This is caused by tauri-1, apparently unresolvable in ubuntu 24.04/mint 22, you need to migrate the app to tauri 2.0..
https://v2.tauri.app/start/migrate/from-tauri-2-beta/
Just run the migration as explained.

In the process, you should have rust > v1.8 installed by now.

Then change in '[config_editor.rs](http://config_editor.rs/)':
"tauri::WindowBuilder::new(app, "main", tauri::WindowUrl::App("index.html".into()))"
to:
"tauri::WebviewWindowBuilder::new(app, "main", tauri::WebviewUrl::App("index.html".into()))"

(Naming changes for the switch from tauri 1 to 2: https://docs.rs/tauri/2.0.0-beta.16/tauri/webview/struct.WebviewWindowBuilder.html)

Then (dending on the migration you have used?) you might get the error:
`failed to build bundler settings: failed to get updater configuration: plugins > updater doesn't exist`

Which is a brand new issue (as I write this text): https://github.com/tauri-apps/tauri/issues/10508
That has a fix that was merged in https://github.com/tauri-apps/tauri/compare/fix/v1-migration and should soon get into main?
But basically, to fix this issue, while not resolved in the migration tool (if by the time you read that, it's resolved, great!).
Else remove the key:
`"createUpdaterArtifacts": "v1Compatible"`
from tauri.conf.json
The conf file should look like this:
```
{
  "build": {
    "beforeDevCommand": "npm run dev",
    "beforeBuildCommand": "npm run build",
    "frontendDist": "../dist",
    "devUrl": "http://localhost:1420/"
  },
  "bundle": {
    "active": true,
    "icon": [
      "../../mouse_actions_logo.png"
    ],
    "targets": "all"
  },
  "productName": "mouse-actions-gui",
  "version": "0.4.5",
  "identifier": "com.soulary.mouse.acions",
  "plugins": {},
  "app": {
    "withGlobalTauri": false,
    "windows": [],
    "security": {
      "csp": null
    }
  }
}
```

Then you might get other errors (from missing dev libs):
run the build with `npm run tauri build -- --verbose` to see the errors in detail and install the missing dependencies

Once all dependencies are installed, the build should end in success :)
And you'll have your package in `config-editor/src-tauri/target/release/bundle/deb/mouse-actions-gui_0.4.5_amd64.deb`

One last note: I opened a blank window for the GUI... somehow, on my version, I need to run the gui with:
WEBKIT_DISABLE_COMPOSITING_MODE=1 mouse_actions_config_editor

Else, I already started configuring some of my favorite gestures, it works perfectly! Thanks

**Comment by m3n3chm0** [comment]
Hello @braikar 

Does it also work for Xubuntu 24.04?

**Reply by braikar** [reply]
Hello,
I don't really know, if easystroke used to work on xubuntu 20.04 or below, then I would say mouseactions can work on 24.04.
Btw. since that time I found even better for my taste. Simply easystroke again, with fixes for 22-24.04!
[https://launchpad.net/~kondor-dani/+archive/ubuntu/ppa-easystroke](https://launchpad.net/~kondor-dani/+archive/ubuntu/ppa-easystroke)

**Reply by mathmul** [reply]
```
sudo add-apt-repository ppa:kondor-dani/ppa-easystroke
sudo apt update
sudo apt install easystroke
``` 
I got it installed, but how does one use to have a 3-finger swipe left/right on the mousepad to work as Ctrl+Alt+Left/Right Arrow, to swipe between Ubuntu workspaces (desktops)?
```
user@x1-extreme ~/Downloads % neofetch
            .-/+oossssoo+/-.               user@x1-extreme
        `:+ssssssssssssssssss+:`           --------------
      -+ssssssssssssssssssyyssss+-         OS: Ubuntu 24.04.2 LTS x86_64
    .ossssssssssssssssssdMMMNysssso.       Host: 20MF000TMZ ThinkPad X1 Extreme
   /ssssssssssshdmmNNmmyNMMMMhssssss/      Kernel: 6.11.0-17-generic
  +ssssssssshmydMMMMMMMNddddyssssssss+     Uptime: 2 hours, 10 mins
 /sssssssshNMMMyhhyyyyhmNMMMNhssssssss/    Packages: 1931 (dpkg), 15 (flatpak), 10 (snap)
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Shell: zsh 5.9
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   Resolution: 3840x2160
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   DE: GNOME 46.0
ossyNMMMNyMMhsssssssssssssshmmmhssssssso   WM: Mutter
+sssshhhyNMMNyssssssssssssyNMMMysssssss+   WM Theme: Adwaita
.ssssssssdMMMNhsssssssssshNMMMdssssssss.   Theme: Yaru [GTK2/3]
 /sssssssshNMMMyhhyyyyhdNMMMNhssssssss/    Icons: Yaru [GTK2/3]
  +sssssssssdmydMMMMMMMMddddyssssssss+     Terminal: WezTerm
   /ssssssssssshdmNNNNmyNMMMMhssssss/      CPU: Intel i7-8750H (12) @ 4.100GHz
    .ossssssssssssssssssdMMMNysssso.       GPU: NVIDIA GeForce GTX 1050 Ti Mobile
      -+sssssssssssssssssyyyssss+-         Memory: 8942MiB / 63829MiB
        `:+ssssssssssssssssss+:`
            .-/+oossssoo+/-.
```

**Comment by m3n3chm0** [comment]
Dear @braikar,
I'll give this a try right now: https://launchpad.net/~kondor-dani/+archive/ubuntu/ppa-easystroke.
Thank you so much in advance!

**Comment by Zefiro** [comment]
Thank you very much for this writeup, @braikar. I managed to come half-way already, until the javascriptcoregtk-4.0 issue.

And as much as I'd like to support new programs - a working, apparently still maintained fork of easystrokes which immediately solves my current issues is too good an opportunity to pass. At least as long as I'm still on X11.

**Reply by braikar** [reply]
Thanks and you're welcome. I reacted totally the same when I found this maintained version of easystroke. :)
